aptos-move/framework/src/natives/aptos_lib.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use move_binary_format::errors::PartialVMResult;
+use move_core_types::{
+    gas_schedule::GasCost,
+    language_storage::{StructTag, TypeTag},
+};
+use move_vm_runtime::native_functions::NativeContext;
+use move_vm_types::{
+    loaded_data::runtime_types::Type, natives::function::NativeResult, values::Value,
+};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, fmt::Write};
+
+/// Return Token types ModuleAddress, ModuleName and StructName
+pub fn type_of(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.len() == 1);
+    debug_assert!(arguments.is_empty());
+
+    let cost = GasCost::new(super::cost::APTOS_LIB_TYPE_OF, 1).total();
+
+    let type_tag = context.type_to_type_tag(&ty_args[0])?;
+    if let TypeTag::Struct(struct_tag) = type_tag {
+        Ok(NativeResult::ok(
+            cost,
+            type_of_internal(&struct_tag).expect("type_of should never fail."),
+        ))
+    } else {
+        Ok(NativeResult::err(
```

Comment:
Are we sure primitives are never needed?

Reply:
That wouldn't work... they aren't structs, we can't parse them.

u64 has no modulename, structname, or address

Reply:
yeah, it can be "builtin::builtin::u64"? if you insist it's three parts..